### PR TITLE
Add Operation CRD and agent daemon operation controller with soft-reboot support

### PR DESCRIPTION
**Pull request by bcho**
## Summary

Replace the temporary ConfigMap-based operation shim with a proper Operation custom resource (`unbounded-kube.io/v1alpha3`) for managing machine operations like soft-reboot.

## Changes

- **Operation CRD**: New cluster-scoped CR with `SoftReboot` and `HardReboot` types, status subresource, owner references for GC, and TTL-based cleanup
- **Agent daemon**: Rewrite operation watcher to watch Operation CRs instead of ConfigMaps; replace `reconciler_softrestart` with generic `reconciler_operation` that dispatches by type
- **kubectl plugin**: Rewrite `soft-reboot` command to create an Operation CR with owner reference and TTL (default 300s), then watch status until completion
- **RBAC**: Replace ConfigMap rules with `operations` and `operations/status` verbs
- **Tests**: 7 operation reconciler tests covering all phases, TTL cleanup, and edge cases

## E2E Validation

Validated on `oc-vm3` (Ubuntu 24.04, Standard_D2as_v7):

```
$ kubectl get machines -o wide
NAME    HOST   PHASE   K8S VERSION   AGE
agent                                14m

$ ./bin/kubectl-unbounded machine soft-reboot agent
  --> Soft-rebooting Machine agent...
  operation         agent-softreboot-1776734492
  --> Operation SoftReboot: agent-softreboot-1776734492 in progress...
  --> Operation SoftReboot: agent-softreboot-1776734492 completed
  ready

$ kubectl get operations -o wide
NAME                          MACHINE   TYPE         PHASE       AGE
agent-softreboot-1776734492   agent     SoftReboot   Completed   4m5s

$ kubectl get operations -o yaml
apiVersion: unbounded-kube.io/v1alpha3
kind: Operation
metadata:
  name: agent-softreboot-1776734492
  ownerReferences:
  - apiVersion: unbounded-kube.io/v1alpha3
    kind: Machine
    name: agent
    uid: d5fc240c-0b2a-4a3f-9468-1d2aabf691bb
spec:
  machineRef: agent
  ttlSecondsAfterFinished: 300
  type: SoftReboot
status:
  completedAt: "2026-04-21T01:21:33Z"
  phase: Completed
  startedAt: "2026-04-21T01:21:32Z"
```

Node returned to `Ready` after soft reboot completed successfully.

### CODE REVIEW

**Comment by bcho** [comment]
```
Events:
  Type     Reason                   Age                    From             Message
  ----     ------                   ----                   ----             -------
  Normal   Starting                 2m4s                   kube-proxy
  Normal   Starting                 15s                    kube-proxy
  Normal   Starting                 11m                    kube-proxy
  Normal   NodeHasNoDiskPressure    6m11s (x8 over 11m)    kubelet          Node oc-vm3 status is now: NodeHasNoDiskPressure
  Normal   NodeHasSufficientPID     6m11s (x8 over 11m)    kubelet          Node oc-vm3 status is now: NodeHasSufficientPID
  Normal   NodeHasSufficientMemory  6m11s (x8 over 11m)    kubelet          Node oc-vm3 status is now: NodeHasSufficientMemory
  Normal   NodeHasSufficientMemory  2m28s (x7 over 2m29s)  kubelet          Node oc-vm3 status is now: NodeHasSufficientMemory
  Normal   NodeHasNoDiskPressure    2m28s (x7 over 2m29s)  kubelet          Node oc-vm3 status is now: NodeHasNoDiskPressure
  Normal   NodeHasSufficientPID     2m28s (x7 over 2m29s)  kubelet          Node oc-vm3 status is now: NodeHasSufficientPID
  Normal   RegisteredNode           2m24s                  node-controller  Node oc-vm3 event: Registered Node oc-vm3 in Controller
  Normal   NodeReady                113s                   kubelet          Node oc-vm3 status is now: NodeReady
  Normal   Starting                 18s                    kubelet          Starting kubelet.
  Normal   NodeHasSufficientMemory  18s (x2 over 18s)      kubelet          Node oc-vm3 status is now: NodeHasSufficientMemory
  Normal   NodeHasNoDiskPressure    18s (x2 over 18s)      kubelet          Node oc-vm3 status is now: NodeHasNoDiskPressure
  Normal   NodeHasSufficientPID     18s (x2 over 18s)      kubelet          Node oc-vm3 status is now: NodeHasSufficientPID
  Normal   NodeAllocatableEnforced  18s                    kubelet          Updated Node Allocatable limit across pods
  Warning  Rebooted                 18s                    kubelet          Node oc-vm3 has been rebooted, boot id: 583dc021-bcda-4f75-a08c-4d7156b8364e
  Warning  InvalidDiskCapacity      18s                    kubelet          invalid capacity 0 on image filesystem
  ```

**Comment by bcho** [comment]
## E2E: Kubernetes Version Upgrade via MachineConfiguration (v1.34.3 -> v1.35.1)

**Cluster:** AKS `bahe-test-nodes` (southcentralus, v1.34.3)
**VM:** oc-vm3 (Ubuntu 24.04, nspawn blue/green)

### Flow

1. `kubectl unbounded config create upgrade-test --kubernetes-version=v1.35.1 --node-labels=...`
2. Controller auto-creates `upgrade-test-v1` MCV
3. `kubectl unbounded config assign upgrade-test agent --version=1`
4. Bump `spec.operations.repaveCounter` to trigger repave
5. Agent resolves MCV, detects version drift, performs full blue/green node update

### Agent Logs

```
[I] daemon starting [machine_cr=agent] [nspawn_machine=kube1] [applied_version=1.34.3]
[I] operation counter drift detected [current_version=1.34.3] [desired_version=v1.35.1] [mcv=upgrade-test-v1]
[I] starting node update [old_machine=kube1] [new_machine=kube2] [old_version=1.34.3] [new_version=1.35.1]
[I] pulling OCI image [image=ghcr.io/azure/agent-ubuntu2404:v20260409]
[I] downloading kubernetes binary [binary=kubelet] [url=https://dl.k8s.io/v1.35.1/bin/linux/amd64/kubelet]
[I] downloading kubernetes binary [binary=kubectl] [url=https://dl.k8s.io/v1.35.1/bin/linux/amd64/kubectl]
[I] downloading kubernetes binary [binary=kube-proxy] [url=https://dl.k8s.io/v1.35.1/bin/linux/amd64/kube-proxy]
[I] stopping machine [machine=kube1]
[I] kubelet is active [machine=kube2]
[I] removing machine rootfs [machine=kube1]
[I] node update completed [active_machine=kube2] [version=1.35.1]
[I] reconciliation completed [new_version=1.35.1] [mcv=upgrade-test-v1]
```

### Controller Logs

```
INFO  creating initial MachineConfigurationVersion  {"name": "upgrade-test", "version": 1}
```

### Machine CR Status (after)

```yaml
status:
  phase: Joining
  message: node update completed
  configuration:
    name: upgrade-test
    version: 1
    versionName: upgrade-test-v1
  conditions:
  - type: NodeUpdated
    status: "True"
    reason: Succeeded
    message: node update completed
  operations:
    repaveCounter: 2
```

### MCV Status

```yaml
spec:
  version: 1
  template:
    kubernetes:
      version: v1.35.1
      nodeLabels:
        kubernetes.azure.com/cluster: bahe-test-nodes
        kubernetes.azure.com/managed: "false"
```

### Node (after)

```
NAME     STATUS   ROLES    AGE     VERSION
oc-vm3   Ready    <none>   3h17m   v1.35.1
```

### Summary

- Total repave time: ~16 seconds (rootfs provision 13.4s + stop/start/cleanup 2.6s)
- Blue/green swap: kube1 (v1.34.3) -> kube2 (v1.35.1)
- Node rejoined cluster as Ready with v1.35.1
- Machine status.configuration correctly records the applied MCV
- RBAC fix: added `machineconfigurationversions` (get/list/watch) to bootstrapper ClusterRole

### kubectl plugin commands tested

```
kubectl unbounded config create upgrade-test --kubernetes-version=v1.35.1 --node-labels=...
kubectl unbounded config get
kubectl unbounded config get upgrade-test
kubectl unbounded config versions upgrade-test
kubectl unbounded config assign upgrade-test agent --version=1
```

**Comment by bcho** [comment]
## E2E: Node Delete as Repave Trigger

Validated the Node-deletion repave signal on `oc-vm3`. The agent now watches the Kubernetes Node object by hostname and triggers a forced repave (bypassing operation counter drift check) when the Node is deleted.

### Resource Setup

**MachineConfiguration** (`upgrade-test`):
```yaml
apiVersion: unbounded-kube.io/v1alpha3
kind: MachineConfiguration
metadata:
  name: upgrade-test
spec:
  priority: 0
  revisionHistoryLimit: 10
  template:
    kubernetes:
      version: v1.35.1
      nodeLabels:
        kubernetes.azure.com/cluster: bahe-test-nodes
        kubernetes.azure.com/managed: "false"
```

**MachineConfigurationVersion v1** (`upgrade-test-v1`) - initial version matching applied:
```yaml
apiVersion: unbounded-kube.io/v1alpha3
kind: MachineConfigurationVersion
metadata:
  name: upgrade-test-v1
  labels:
    unbounded-kube.io/configuration: upgrade-test
spec:
  version: 1
  template:
    kubernetes:
      version: v1.35.1
      nodeLabels:
        kubernetes.azure.com/cluster: bahe-test-nodes
        kubernetes.azure.com/managed: "false"
```

**MachineConfigurationVersion v2** (`upgrade-test-v2`) - target version for repave:
```yaml
apiVersion: unbounded-kube.io/v1alpha3
kind: MachineConfigurationVersion
metadata:
  name: upgrade-test-v2
  labels:
    unbounded-kube.io/configuration: upgrade-test
spec:
  version: 2
  template:
    kubernetes:
      version: v1.34.3
      nodeLabels:
        kubernetes.azure.com/cluster: bahe-test-nodes
        kubernetes.azure.com/managed: "false"
```

**Machine CR** (`agent`) - assigned to MCV v2:
```yaml
apiVersion: unbounded-kube.io/v1alpha3
kind: Machine
metadata:
  name: agent
spec:
  configurationRef:
    name: upgrade-test
    version: 2
  kubernetes:
    bootstrapTokenRef:
      name: bootstrap-token-ftbv20
    nodeLabels:
      kubernetes.azure.com/cluster: bahe-test-nodes
      kubernetes.azure.com/managed: "false"
```

### Test 1: Node delete with no config drift

Assigned MCV v1 (v1.35.1, same as applied), deleted Node `oc-vm3`:

```
22:44:17 [I] Node deleted, enqueuing repave [watcher=node] [node=oc-vm3]
22:44:17 [I] Node deleted, forcing repave [action=NodeDeleted] [source=oc-vm3]
                [current_version=1.35.1] [desired_version=v1.35.1] [mcv=upgrade-test-v1]
22:44:17 [I] no config drift detected, skipping node update [action=NodeDeleted] [source=oc-vm3]
22:44:17 [I] reconciliation completed [new_version=1.35.1] [mcv=upgrade-test-v1] [action=NodeDeleted]
```

Result: Agent detected Node deletion, bypassed operation counter check (forceRepave=true), but `updateNode` correctly found no config drift and skipped the expensive repave. No unnecessary work.

### Test 2: Node delete with version change (v1.35.1 -> v1.34.3)

Assigned MCV v2 (v1.34.3), deleted Node `oc-vm3`:

```
22:49:24 [I] Node deleted, enqueuing repave [watcher=node] [node=oc-vm3]
22:49:24 [I] Node deleted, forcing repave [action=NodeDeleted] [source=oc-vm3]
                [current_version=1.35.1] [desired_version=v1.34.3] [mcv=upgrade-test-v2]
22:49:24 [I] starting node update [old_machine=kube2] [new_machine=kube1]
                [old_version=1.35.1] [new_version=1.34.3]
22:49:24 [I] pulling OCI image [image=ghcr.io/azure/agent-ubuntu2404:v20260409]
22:49:26 [I] OCI image extraction complete                              (2.0s)
22:49:28 [I] downloaded kube binaries (v1.34.3)                         (2.0s)
22:49:30 [I] stopping machine [machine=kube2]
22:49:32 [I] [stop-node] completed                                      (1.9s)
22:49:32 [I] [start-nspawn-machine] started (kube1)
22:49:33 [I] kubelet is active [machine=kube1]
22:49:33 [I] removing machine rootfs [machine=kube2]
22:49:34 [I] node update completed [active_machine=kube1] [version=1.34.3]
22:49:34 [I] reconciliation completed [new_version=1.34.3] [mcv=upgrade-test-v2]
```

Total repave time: **~10 seconds** (22:49:24 -> 22:49:34). Blue/green: kube2 -> kube1.

### Post-repave state

**Machine status** after repave:
```yaml
status:
  phase: Joining
  message: node update completed
  configuration:
    name: upgrade-test
    version: 2
    versionName: upgrade-test-v2
  conditions:
  - type: NodeUpdated
    status: "True"
    reason: Succeeded
    message: node update completed
    observedGeneration: 12
```

**Node** re-registered with new version:
```
NAME     STATUS   ROLES    AGE   VERSION   INTERNAL-IP
oc-vm3   Ready    <none>   9m    v1.34.3   10.1.0.4
```

### Agent watchers

On startup, agent now runs three concurrent watch loops:
```
22:43:30 [I] daemon starting [machine_cr=agent] [nspawn_machine=kube2] [applied_version=1.35.1]
22:43:30 [I] Node watch starting [watcher=node] [node=oc-vm3]
22:43:30 [I] watching Node [name=oc-vm3]
22:43:30 [I] watching MachineOperation CRs [machineRef=agent]
22:43:30 [I] watching Machine CR [name=agent]
```

### RBAC

Added to `07-bootstrapper-rbac.yaml`:
```yaml
- apiGroups: [""]
  resources: ["nodes"]
  verbs: ["get", "list", "watch"]
```

### New files

- `cmd/agent/internal/daemon/nodewatch.go` - Node watch by hostname, enqueues `ActionNodeDeleted` on delete
- `cmd/agent/internal/daemon/nodewatch_test.go` - 2 tests (delete enqueue, hostname error)
- Updated `reconciler.go` dispatch for `ActionNodeDeleted`
- Updated `reconcileUpdateMachine` with `forceRepave bool` parameter (skips operation counter drift check)
- 2 new reconciler tests (`ForceRepaveSkipsDriftCheck`, `NoDrift_NormalPath`)

**Comment by bcho** [comment]
## kubectl-unbounded config commands

The following `kubectl unbounded config` subcommands were added to manage the MachineConfiguration workflow:

### `kubectl unbounded config create`

Creates a MachineConfiguration resource. The machina controller automatically creates MCV v1 from the provided settings.

```
kubectl unbounded config create my-config \
  --kubernetes-version=v1.34.0 \
  --agent-image=ghcr.io/azure/unbounded-agent:latest \
  --node-labels=env=prod,tier=worker \
  --taints=dedicated=gpu:NoSchedule \
  --update-strategy=OnDelete \
  --priority=10
```

### `kubectl unbounded config get`

Lists MachineConfigurations, or shows details for a specific one:

```
kubectl unbounded config get
kubectl unbounded config get my-config
```

### `kubectl unbounded config versions`

Lists all MachineConfigurationVersions for a MachineConfiguration, ordered by version number:

```
kubectl unbounded config versions my-config
```

### `kubectl unbounded config assign`

Sets `spec.configurationRef` on a Machine to reference a MachineConfiguration. Optionally pins a specific version:

```
kubectl unbounded config assign my-config my-machine
kubectl unbounded config assign my-config my-machine --version=3
```

### E2E workflow example (what was used in the Node-delete test above)

```bash
# 1. Create the MachineConfiguration (or apply manually if machina controller isn't updated)
kubectl unbounded config create upgrade-test --kubernetes-version=v1.35.1 \
  --node-labels=kubernetes.azure.com/cluster=bahe-test-nodes,kubernetes.azure.com/managed=false

# 2. Assign MCV v1 to Machine
kubectl unbounded config assign upgrade-test agent --version=1

# 3. Create MCV v2 with different version (manually, since machina controller wasn't updated)
# In production the controller auto-creates MCVs when the MC spec changes.

# 4. Assign MCV v2 to Machine
kubectl unbounded config assign upgrade-test agent --version=2

# 5. Trigger repave via Node deletion (OnDelete strategy)
kubectl delete node oc-vm3
# Agent detects deletion -> resolves MCV v2 -> repaves kube2->kube1 with v1.34.3
```

**Comment by bcho** [comment]
## E2E: Soft Reboot via MachineOperation

Validated `kubectl unbounded machine soft-reboot` on `oc-vm3` (currently running v1.34.3 on kube1 after the Node-delete repave above).

### Command

```bash
kubectl unbounded machine soft-reboot agent
```

Output:
```
  --> Soft-rebooting Machine agent...
  operation         agent-reboot-1776812997

  --> Operation Reboot: agent-reboot-1776812997 in progress...
  --> Operation Reboot: agent-reboot-1776812997 completed

  ready
```

### MachineOperation CR

```yaml
apiVersion: unbounded-kube.io/v1alpha3
kind: MachineOperation
metadata:
  name: agent-reboot-1776812997
  labels:
    unbounded-kube.io/machine: agent
  ownerReferences:
  - apiVersion: unbounded-kube.io/v1alpha3
    kind: Machine
    name: agent
    uid: d81834c0-df27-4a32-a98a-0ee1b209fc9c
spec:
  machineRef: agent
  operationName: Reboot
  ttlSecondsAfterFinished: 300
status:
  phase: Complete
  startedAt: "2026-04-21T23:09:57Z"
  completedAt: "2026-04-21T23:09:58Z"
```

### Agent logs

```
23:09:57 [I] executing operation [op_name=Reboot] [source=agent-reboot-1776812997]
23:09:57 [I] soft restart: pre-stopping services in machine [machine=kube1]
23:09:57 [I] soft restart: restarting nspawn service [service=systemd-nspawn@kube1.service]
23:09:58 [I] kubelet is active [machine=kube1]
23:09:58 [I] soft restart: completed [machine=kube1]
23:09:58 [I] operation completed [op_name=Reboot] [source=agent-reboot-1776812997]
```

Total soft-reboot time: **~1 second** (23:09:57 -> 23:09:58). The nspawn container was restarted in place without reprovisioning rootfs. The MachineOperation CR auto-deletes after 300s (TTL).

File: deploy/machina/crd/unbounded-cloud.io_machineoperations.yaml [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+              machineRef:
+                description: MachineRef is the name of the Machine CR this operation
+                  targets.
+                minLength: 1
+                type: string
+              operationName:
+                description: OperationName is the operation to perform on the target
+                  machine.
+                enum:
+                - Reboot
```
Comment (phealy) [inline deploy/machina/crd/unbounded-cloud.io_machineoperations.yaml]: The CRD definition is using the values I suggested (Reboot/PowerCycle) but I think the POC is using SoftRestart and HardRestart). I personally prefer reboot/powercycle as I think they're clearer, but either way can we be consistent?

**Comment by plombardi89** [comment]
I feel like the `unbounded config` should move to `unbounded machine config` or if we want a top-level command `unbounded machine-config` to express the relationship better. The current relationship `unbounded config` makes me think it's a high-level system configuration.

File: deploy/machina/crd/unbounded-cloud.io_machineoperations.yaml (hunk 1 above)
Comment (plombardi89) [reply deploy/machina/crd/unbounded-cloud.io_machineoperations.yaml]: I kind of prefer the Soft/Hard Restart (or Reboot) myself. PowerCycle implies something very specific which may or may not be happening depending on the provider implementation below us.

Comment (phealy) [reply deploy/machina/crd/unbounded-cloud.io_machineoperations.yaml]: OK, I'm certainly open to what others think is clearer!

**Comment by bcho** [comment]
> I feel like the `unbounded config` should move to `unbounded machine config` or if we want a top-level command `unbounded machine-config` to express the relationship better. The current relationship `unbounded config` makes me think it's a high-level system configuration.

But the config is actually assigned to multiple machines, not sure if putting it as "sub command" under machine causes confusion.  @phealy wdyt?

**Comment by bcho** [comment]
## E2E Test Results — `eacfce0` (PowerCycle→HardReboot / softRestart→softReboot rename)

All 3 tests passed on **oc-vm3** with the latest agent binary.

| Test | Result | Duration | Details |
|------|--------|----------|---------|
| Node-delete repave | **PASS** | ~10s | v1.34.3 → v1.35.1, kube1 → kube2, MCV v2 → v3 |
| HardReboot operation | **PASS** (expected fail) | <1s | Correctly rejected: "handled by machina controller, not the agent" |
| Soft Reboot operation | **PASS** | ~1s | kube2 soft rebooted, node re-registered Ready |

### Command Flows

#### 1. MachineConfiguration + MCV lifecycle

```bash
# Create a MachineConfiguration
kubectl apply -f - <<YAML
apiVersion: unbounded-kube.io/v1alpha3
kind: MachineConfiguration
metadata:
  name: upgrade-test
spec:
  template:
    kubernetes:
      version: "v1.34.3"
      nodeLabels:
        kubernetes.azure.com/cluster: bahe-test-nodes
YAML

# Create versioned snapshots (MCVs)
kubectl apply -f - <<YAML
apiVersion: unbounded-kube.io/v1alpha3
kind: MachineConfigurationVersion
metadata:
  name: upgrade-test-v1
  labels:
    unbounded-kube.io/configuration: upgrade-test
spec:
  version: 1
  template:
    kubernetes:
      version: "v1.34.3"
      nodeLabels:
        kubernetes.azure.com/cluster: bahe-test-nodes
YAML

# List configurations and versions
kubectl get machineconfigurations
kubectl get machineconfigurationversions
```

#### 2. Assign a config version to a Machine

```bash
# Point Machine to a specific MCV
kubectl patch machine agent --type=merge \
  -p '{"spec":{"configurationRef":{"name":"upgrade-test","version":3}}}'

# Verify assignment
kubectl get machine agent -o jsonpath='{.spec.configurationRef}'
kubectl get machine agent -o jsonpath='{.status.configuration}'
```

#### 3. Trigger repave via Node delete (OnDelete strategy)

```bash
# Agent detects drift but waits for Node delete signal.
# Delete the Node to trigger repave:
kubectl delete node oc-vm3

# Agent detects deletion -> repaves to target MCV version.
# Node re-registers automatically after repave (~10-15s).
kubectl get node oc-vm3
```

#### 4. MachineOperations

```bash
# Soft reboot (handled by in-VM agent)
kubectl apply -f - <<YAML
apiVersion: unbounded-kube.io/v1alpha3
kind: MachineOperation
metadata:
  name: reboot-1
  labels:
    unbounded-kube.io/machine: agent
spec:
  machineRef: agent
  operationName: Reboot
YAML

# HardReboot (rejected by agent - handled by machina controller)
kubectl apply -f - <<YAML
apiVersion: unbounded-kube.io/v1alpha3
kind: MachineOperation
metadata:
  name: hardreboot-1
  labels:
    unbounded-kube.io/machine: agent
spec:
  machineRef: agent
  operationName: HardReboot
YAML

# Check operation status
kubectl get machineoperations -o wide
```

#### 5. kubectl unbounded plugin commands

```bash
# Create a configuration
kubectl unbounded config create my-config --k8s-version v1.35.1

# List configurations
kubectl unbounded config get

# List versions for a configuration
kubectl unbounded config versions upgrade-test

# Assign a version to a machine
kubectl unbounded config assign upgrade-test --version 3 --machine agent
```

**Comment by bcho** [comment]
closing out this one with the real impl
